File: src/main/java/com/dyonovan/neotech/client/modelfactory/models/ModelPipe.java
package com.dyonovan.neotech.client.modelfactory.models;

import com.dyonovan.neotech.collections.DummyState;
import com.dyonovan.neotech.common.blocks.pipe.BlockPipe;
import com.dyonovan.neotech.lib.Constants;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.model.*;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.resources.model.IBakedModel;
import net.minecraft.client.resources.model.ModelRotation;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.client.model.ISmartBlockModel;
import net.minecraftforge.client.model.ISmartItemModel;

import javax.vecmath.Vector3f;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings( "deprecation" )
public class ModelPipe implements ISmartBlockModel, ISmartItemModel {
    protected TextureAtlasSprite backGround = null;
    protected TextureAtlasSprite foreGround = null;
    protected BlockPipe pipe;
    protected BlockPos pos;
    protected IBlockAccess worldObj;
    static FaceBakery faceBakery = new FaceBakery();
    boolean[] extensions = new boolean[6];
    protected float min;
    protected float max;

    public ModelPipe(DummyState state) {
        for(EnumFacing face : EnumFacing.values()) {
            extensions[face.ordinal()] = ((BlockPipe)state.block).isCableConnected(state.blockAccess, state.pos.offset(face), face);
        }

        backGround = Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite(((BlockPipe)state.block).getBackgroundTexture());
        foreGround = Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite(Constants.MODID + ":blocks/" + state.block.getName());

        min = 8.0F - ((BlockPipe)state.block).getWidth();
        max = 8.0F +  ((BlockPipe)state.block).getWidth();

        pipe = (BlockPipe) state.block;
        worldObj = state.blockAccess;
        pos = state.pos;
    }

    public ModelPipe(BlockPipe pipe) {
        backGround = Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite(pipe.getBackgroundTexture());
        foreGround = Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite(Constants.MODID + ":blocks/" + pipe.getName());

        min = 8.0F - pipe.getWidth();
        max = 8.0F + pipe.getWidth();
    }

    public List<BakedQuad> getFaceQuads(EnumFacing facing) {
        return new ArrayList<>();
    }

    private void drawPipeConnection(ModelRotation modelRot, List<BakedQuad> list) {
        BlockPartFace face = new BlockPartFace(null, 0, "", new BlockFaceUV(new float[]{0.0F, 0.0F, 16.0F, 16.0F}, 0));
        boolean scale = true;
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, max, max), new Vector3f(max, max, 16.0F), face, backGround, EnumFacing.UP, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, max), new Vector3f(max, min, 16.0F), face, backGround, EnumFacing.DOWN, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, max), new Vector3f(min, max, 16.0F), face, backGround, EnumFacing.WEST, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(max, min, max), new Vector3f(max, max, 16.0F), face, backGround, EnumFacing.EAST, modelRot, null, scale, true));
    }

    public List<BakedQuad> getGeneralQuads() {
        ArrayList list = new ArrayList();
        BlockFaceUV uv = new BlockFaceUV(new float[]{0.0F, 0.0F, 16.0F, 16.0F}, 0);
        BlockPartFace face = new BlockPartFace(null, 0, "", uv);

        if(this.extensions[EnumFacing.DOWN.ordinal()]) {
            this.drawPipeConnection(ModelRotation.X270_Y0, list);
        }

        if(this.extensions[EnumFacing.UP.ordinal()]) {
            this.drawPipeConnection(ModelRotation.X90_Y0, list);
        }

        if(this.extensions[EnumFacing.NORTH.ordinal()]) {
            this.drawPipeConnection(ModelRotation.X180_Y0, list);
        }

        if(this.extensions[EnumFacing.SOUTH.ordinal()]) {
            this.drawPipeConnection(ModelRotation.X0_Y0, list);
        }

        if(this.extensions[EnumFacing.WEST.ordinal()]) {
            this.drawPipeConnection(ModelRotation.X0_Y90, list);
        }

        if(this.extensions[EnumFacing.EAST.ordinal()]) {
            this.drawPipeConnection(ModelRotation.X0_Y270, list);
        }

        ModelRotation modelRot = ModelRotation.X0_Y0;
        boolean scale = true;
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, backGround, EnumFacing.DOWN, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, backGround, EnumFacing.UP, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, backGround, EnumFacing.NORTH, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, backGround, EnumFacing.SOUTH, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, backGround, EnumFacing.EAST, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, backGround, EnumFacing.WEST, modelRot, null, scale, true));

        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, foreGround, EnumFacing.DOWN, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, foreGround, EnumFacing.UP, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, foreGround, EnumFacing.NORTH, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, foreGround, EnumFacing.SOUTH, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, foreGround, EnumFacing.EAST, modelRot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, foreGround, EnumFacing.WEST, modelRot, null, scale, true));

        if(worldObj != null && pipe != null) {
            pipe.drawExtras(list, faceBakery, worldObj, pos);
        }

        return list;
    }

    public boolean isAmbientOcclusion() {
        return false;
    }

    public boolean isGui3d() {
        return true;
    }

    public boolean isBuiltInRenderer() {
        return false;
    }

    public TextureAtlasSprite getTexture() {
        return backGround;
    }

    public static final ItemTransformVec3f LittleBigger = new ItemTransformVec3f(new Vector3f(), new Vector3f(), new Vector3f(1.5F, 1.5F, 1.5F));
    public static final ItemTransformVec3f MovedUp = new ItemTransformVec3f(new Vector3f(), new Vector3f(-0.05F, 0.0F, -0.15F), new Vector3f(1.0F, 1.0F, 1.0F));

    public ItemCameraTransforms getItemCameraTransforms() {
        return new ItemCameraTransforms(MovedUp, LittleBigger, LittleBigger, LittleBigger);
    }

    public IBakedModel handleBlockState(IBlockState state) {
        return state instanceof DummyState ? new ModelPipe((DummyState)state) : null;
    }

    @Override
    public IBakedModel handleItemState(ItemStack stack) {
        return new ModelPipe((BlockPipe) Block.getBlockFromItem(stack.getItem()));
    }
}

File: src/main/scala/com/dyonovan/neotech/client/modelfactory/models/ModelPipe.scala
package com.dyonovan.neotech.client.modelfactory.models

import java.util
import javax.vecmath.Vector3f

import com.dyonovan.neotech.client.modelfactory.ModelFactory
import com.dyonovan.neotech.managers.BlockManager
import com.dyonovan.neotech.pipes.blocks.BlockPipe
import com.dyonovan.neotech.pipes.types.SimplePipe
import com.teambr.bookshelf.common.blocks.properties.TileAwareState
import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.block.model._
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.resources.model.{ ModelRotation, IBakedModel }
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraftforge.client.model.{ ISmartBlockModel, ISmartItemModel }

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis pauljoda
 * @since August 14, 2015
 */
class ModelPipe extends ISmartBlockModel with ISmartItemModel {
    val faceBakery = new FaceBakery
    val uv = new BlockFaceUV(Array[Float](0.0F, 0.0F, 16.0F, 16.0F), 0)
    val face = new BlockPartFace(null, 0, "", uv)
    var blockPipe : BlockPipe = null
    var pipeTile : SimplePipe = null
    var min = 4.0F
    var max = 12.0F

    def this(block : BlockPipe, tile: SimplePipe) {
        this()
        blockPipe = block
        pipeTile = tile
    }


    override def getFaceQuads(facing : EnumFacing) : util.List[_] = new util.ArrayList[Nothing]()

    override def getGeneralQuads : util.List[_] = {
        val list = new util.ArrayList[BakedQuad]()

        for(dir <- EnumFacing.values()) {
            //Draw the center bit, front and back
            list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, min), new Vector3f(max, max, max), face, getTexture, dir, ModelRotation.X0_Y0, null, true, true))
            list.add(faceBakery.makeBakedQuad(new Vector3f(max, max, max), new Vector3f(min, min, min), face, getTexture, dir.getOpposite, ModelRotation.X0_Y0, null, true, true))
        }

        if(pipeTile != null) { //Exists in world
            if (pipeTile.canConnect(EnumFacing.DOWN)) {
                drawPipeConnection(ModelRotation.X270_Y0, list)
                if(getExtraTexture != null && pipeTile.isSpecialConnection(EnumFacing.DOWN))
                    drawPipeExtras(ModelRotation.X270_Y0, list)
            }
            if (pipeTile.canConnect(EnumFacing.UP)) {
                drawPipeConnection(ModelRotation.X90_Y0, list)
                if(getExtraTexture != null && pipeTile.isSpecialConnection(EnumFacing.UP))
                    drawPipeExtras(ModelRotation.X90_Y0, list)
            }
            if (pipeTile.canConnect(EnumFacing.NORTH)) {
                drawPipeConnection(ModelRotation.X180_Y0, list)
                if(getExtraTexture != null && pipeTile.isSpecialConnection(EnumFacing.NORTH))
                    drawPipeExtras(ModelRotation.X180_Y0, list)
            }
            if (pipeTile.canConnect(EnumFacing.SOUTH)) {
                drawPipeConnection(ModelRotation.X0_Y0, list)
                if(getExtraTexture != null && pipeTile.isSpecialConnection(EnumFacing.SOUTH))
                    drawPipeExtras(ModelRotation.X0_Y0, list)
            }
            if (pipeTile.canConnect(EnumFacing.WEST)) {
                drawPipeConnection(ModelRotation.X0_Y90, list)
                if(getExtraTexture != null && pipeTile.isSpecialConnection(EnumFacing.WEST))
                    drawPipeExtras(ModelRotation.X0_Y90, list)
            }
            if (pipeTile.canConnect(EnumFacing.EAST)) {
                drawPipeConnection(ModelRotation.X0_Y270, list)
                if(getExtraTexture != null && pipeTile.isSpecialConnection(EnumFacing.EAST))
                    drawPipeExtras(ModelRotation.X0_Y270, list)
            }
        } else { //Is an item
            drawPipeConnection(ModelRotation.X270_Y0, list)
            drawPipeConnection(ModelRotation.X90_Y0, list)
            if(getExtraTexture != null)
                drawPipeExtras(ModelRotation.X270_Y0, list)
        }

        list
    }

    def drawPipeConnection(modelRot : ModelRotation, list : util.ArrayList[BakedQuad]) {
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, max, max), new Vector3f(max, max, 16.0F), face, getTexture, EnumFacing.UP, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, max), new Vector3f(max, min, 16.0F), face, getTexture, EnumFacing.DOWN, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, min, max), new Vector3f(min, max, 16.0F), face, getTexture, EnumFacing.WEST, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(max, min, max), new Vector3f(max, max, 16.0F), face, getTexture, EnumFacing.EAST, modelRot, null, true, true))

        //Enable the back to be seen
        list.add(faceBakery.makeBakedQuad(new Vector3f(max, max, 16.0F), new Vector3f(min, max, max), face, getTexture, EnumFacing.DOWN, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(max, min, 16.0F), new Vector3f(min, min, max), face, getTexture, EnumFacing.UP, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(min, max, 16.0F), new Vector3f(min, min, max), face, getTexture, EnumFacing.EAST, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(max, max, 16.0F), new Vector3f(max, min, max), face, getTexture, EnumFacing.WEST, modelRot, null, true, true))
    }

    def drawPipeExtras(modelRot : ModelRotation, list : util.ArrayList[BakedQuad]) {
        list.add(faceBakery.makeBakedQuad(new Vector3f(min - 1, max + 1, max), new Vector3f(max + 1, max + 1, 16.0F), face, getExtraTexture, EnumFacing.UP, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(min - 1, min - 1, max), new Vector3f(max + 1, min - 1, 16.0F), face, getExtraTexture, EnumFacing.DOWN, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(min - 1, min - 1, max), new Vector3f(min - 1, max + 1, 16.0F), face, getExtraTexture, EnumFacing.WEST, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(max + 1, min - 1, max), new Vector3f(max + 1, max + 1, 16.0F), face, getExtraTexture, EnumFacing.EAST, modelRot, null, true, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(min - 1, min - 1, max - 0.001F), new Vector3f(max + 1, max + 1, max - 0.001F), face, getExtraTexture, EnumFacing.NORTH, modelRot, null, true, true))
    }

    override def isAmbientOcclusion: Boolean = true

    override def isGui3d: Boolean = true

    override def isBuiltInRenderer: Boolean = false

    val MovedUp = new ItemTransformVec3f(new Vector3f(0.0F, 0.0F, 0.0F), new Vector3f(-0.05F, 0.05F, -0.15F), new Vector3f(-0.5F, -0.5F, -0.5F))
    override def getItemCameraTransforms : ItemCameraTransforms = {
        new ItemCameraTransforms(MovedUp, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT)
    }

    override def getTexture : TextureAtlasSprite =  {
        blockPipe match {
            case BlockManager.pipeItemSource =>
                ModelFactory.BASIC_ITEM_SOURCE
            case BlockManager.pipeItemSink =>
                ModelFactory.BASIC_ITEM_SINK
            case BlockManager.pipeEnergySource =>
                ModelFactory.BASIC_ENERGY_SOURCE
            case BlockManager.pipeEnergySink =>
                ModelFactory.BASIC_ENERGY_SINK
            case BlockManager.pipeFluidSource =>
                ModelFactory.BASIC_FLUID_SOURCE
            case BlockManager.pipeFluidSink =>
                ModelFactory.BASIC_FLUID_SINK
            case _ =>
                ModelFactory.STRUCTURE_PIPE
        }
    }

    def getExtraTexture : TextureAtlasSprite = {
        blockPipe match {
            case BlockManager.pipeItemSource =>
                ModelFactory.BASIC_ITEM_SOURCE_EXTRAS
            case BlockManager.pipeItemSink =>
                ModelFactory.BASIC_ITEM_SINK_EXTRAS
            case BlockManager.pipeEnergySource =>
                ModelFactory.BASIC_ENERGY_SOURCE_EXTRAS
            case BlockManager.pipeEnergySink =>
                ModelFactory.BASIC_ENERGY_SINK_EXTRAS
            case BlockManager.pipeFluidSource =>
                ModelFactory.BASIC_FLUID_SOURCE_EXTRAS
            case BlockManager.pipeFluidSink =>
                ModelFactory.BASIC_FLUID_SINK_EXTRAS
            case _ =>
                null
        }
    }

    override def handleBlockState(state : IBlockState) : IBakedModel =  {
        state match {
            case tileAware : TileAwareState => new ModelPipe(state.getBlock.asInstanceOf[BlockPipe], tileAware.tile.asInstanceOf[SimplePipe])
            case _ =>  new ModelPipe

        }
    }

    override def handleItemState(stack : ItemStack) : IBakedModel = new ModelPipe(Block.getBlockFromItem(stack.getItem).asInstanceOf[BlockPipe], null)
}
